bootstrap/bootstrap.go

```diff
@@ -946,9 +951,63 @@ func hasGitSubmodules(sh *shell.Shell) bool {
 	return fileExists(filepath.Join(sh.Getwd(), ".gitmodules"))
 }
 
+func (b *Bootstrap) referenceRepository() (string, error) {
+	path := os.Getenv(`BUILDKITE_REPO_MIRROR_PATH`)
+	if path == "" {
+		path = filepath.Join(b.Config.ReposPath, dirForRepository(b.Repository))
+	}
+
+	// Create the base dir if it doesn't exist
+	if baseDir := filepath.Dir(path); !fileExists(baseDir) {
+		b.shell.Commentf("Creating \"%s\"", baseDir)
+		if err := os.MkdirAll(baseDir, 0777); err != nil {
+			return "", err
+		}
+	}
+
+	// Try and lock the repository dir to prevent concurrent clones
+	repoDirLock, err := b.shell.LockFile(path+".lock", time.Minute*5)
+	if err != nil {
+		return "", err
+	}
+	defer repoDirLock.Unlock()
+
+	if !fileExists(path) {
+		b.shell.Commentf("Cloning a mirror of the repository to %s", path)
+		if err := gitCloneMirror(b.shell, b.GitCloneFlags, b.Repository, path); err != nil {
+			return "", err
+		}
+	} else {
```

Comment:
Do you need to worry about the case where the directory already exists, but it's not actually a valid clone?  EG - if someone accidentally/maliciously sets the mirror dir env var to a value that makes the agent attempt to overwrite unrelated contents on the machine's drive?

Reply:
That's not really a threat model that we protect against, you could already do that by setting `BUILDKITE_BUILD_CHECKOUT_PATH` couldn't you?